Refactor(game-client): added detection with hash-based identification and GeneralsOnline support

undead2146 opened the pull request:
## Overview
Refactoring of game client detection and manifest generation systems to support hash-based executable identification, GeneralsOnline clients, and improved manifest ID structure.

* Merge after #164 
## Key Changes

### Game Client Detection
- Introduced `GameClientHashRegistry` implementing `IGameClientHashRegistry` for managing known executable SHA256 hashes with version mappings
- Created `GameClientInfo` class containing `GameType`, `Version`, `Publisher`, `Description`, and `IsOfficial` fields
- Added `GameClientConstants` with executable names, required DLLs, and config files for both vanilla and GeneralsOnline clients
- Refactored `GameClientDetector` to use hash-based detection instead of filename patterns
- Added `DetectGeneralsOnlineClientsAsync()` to detect GeneralsOnline 30Hz/60Hz variants
- Renamed interface methods: `DetectClientsFromInstallationsAsync` → `DetectGameClientsFromInstallationsAsync`, `ScanDirectoryForClientsAsync` → `ScanDirectoryForGameClientsAsync`, `ValidateClientAsync` → `ValidateGameClientAsync`

### Manifest ID System
- Modified `ManifestIdGenerator.GeneratePublisherContentId()` to accept `ContentType` parameter and optional suffix
- Changed `ManifestIdGenerator.GenerateGameInstallationId()` to accept `object?` for version (supports both int and string like "1.04")
- Added `NormalizeVersionString()` to convert version formats (e.g., "1.04" → "104")
- Updated `ManifestIdValidator` with separate regex patterns for installation vs publisher content IDs
- Modified `IManifestIdService.GeneratePublisherContentId()` to require `ContentType` parameter
- Changed `IManifestIdService.GenerateGameInstallationId()` to accept `object?` for version parameter

### Game Installation Management
- Added `GameInstallation.Id` property (no longer auto-generated GUID)
- Renamed `GameInstallation.AvailableClients` → `GameInstallation.AvailableGameClients`
- Added `GameInstallation.SetPaths()` to explicitly set Generals/ZeroHour paths with validation
- Added `GameInstallation.PopulateGameClients()` to associate detected clients with installation
- Improved `GameInstallation.Fetch()` with executable validation (`generals.exe`, `game.exe`)
- Added `GameInstallation.HasValidExecutable()` private method
- Updated `IGameInstallation` interface with new methods and properties
- Implemented interface changes in `SteamInstallation`, `WineInstallation`, and `EaAppInstallation` across platforms

### Manifest Generation
- Refactored `ManifestGenerationService.CreateGameInstallationManifestAsync()` with recursive directory scanning and skip logic for non-game directories (RedistInstallers, Manuals, launcher, .GenLauncherFolder)
- Added `ManifestGenerationService.CreateGameClientManifestAsync()` overload accepting installation path, game type, client details
- Added `ManifestGenerationService.CreateGeneralsOnlineClientManifestAsync()` with Zero Hour dependency
- Implemented selective file hashing in `ContentManifestBuilder`: only hash files when `ContentSourceType != GameInstallation` or `isExecutable == true`
- Added `ContentManifestBuilder.AddGameInstallationFileAsync()` extension method
- Modified `IContentManifestBuilder.WithBasicInfo()` to accept `object?` for manifestVersion
- Added `IContentManifestBuilder.WithBasicInfo()` overload accepting `PublisherInfo`
- Updated `IContentManifestBuilder.WithPublisher()` to include `publisherType` parameter

### Dependencies and Metadata
- Added `ContentDependency` fields: `PublisherType`, `StrictPublisher`, `ExactVersion`, `IsOptional`, `FallbackDependency`
- Added `PublisherInfo` fields: `PublisherType`, `UpdateApiEndpoint`, `ContentIndexUrl`, `UpdateCheckIntervalHours`, `SupportsIncrementalUpdates`, `AuthenticationMethod`
- Created `PublisherTypeConstants` with constants (EaApp, Steam, Retail, GeneralsOnline, etc.) and `FromInstallationType()` mapping
- Added `InstallationExtensions.ToIdentifierString()` for standardized installation type strings

### Process Management
- Added `IGameProcessManager.ProcessExited` event with `GameProcessExitedEventArgs` containing `ProcessId`, `ExitCode`, `ExitTime`
- Implemented `GameProcessManager.OnProcessExited()` handler to fire event and remove from managed processes dictionary

### Constants
- Updated `ManifestConstants.DefaultContentDependencyId` from `"1.0.test.dependency.content"` → `"1.0.genhub.content.defaultdependency"`
- Added `ManifestConstants.GeneralsManifestVersion = "1.08"` and `ManifestConstants.ZeroHourManifestVersion = "1.04"`

ViTeXFTW commented on GenHub/GenHub.Core/Interfaces/Manifest/IManifestGenerationService.cs:
```diff
@@ -16,7 +16,7 @@ public interface IManifestGenerationService
     /// <param name="manifestVersion">The manifest version (e.g., 1, 2, 20). Defaults to 0 for first version.</param>
     /// <returns>A <see cref="Task"/> that returns a configured manifest builder.</returns>
-    Task<IContentManifestBuilder> CreateGameInstallationManifestAsync(string gameInstallationPath, GameType gameType, GameInstallationType installationType, int manifestVersion = 0);
+    Task<IContentManifestBuilder> CreateGameInstallationManifestAsync(string gameInstallationPath, GameType gameType, GameInstallationType installationType, object? manifestVersion = null);
```
Not sure why we would use an object here? What stops us from using integer?

ViTeXFTW commented on GenHub/GenHub.Core/Models/Manifest/ManifestIdValidator.cs:
```diff
@@ -10,17 +11,24 @@ namespace GenHub.Core.Models.Manifest;
+
+    // Allow simple test-friendly ids like 'test-id' or 'simple.id' (alphanumeric with dashes, max 4 segments)
+    private static readonly Regex SimpleIdRegex =
+        new(@"^[a-zA-Z0-9\-]+(\.[a-zA-Z0-9\-]+){0,3}$", RegexOptions.Compiled);
```
`-` doesn't need `\`

undead2146 commented:
@ViTeXFTW I resolved the code changes you requested.
Regarding why I used object, this is because technically Strings like `"1.04"` and Integers `0` are allowed but eventually they all get normalised to integers like `104`  in order to comply with the 5-segment Manifest-ID. 

So I removed the object paramater and added an overload to the interface so now both strings and integers are accepted.

RMWMStevens commented on GenHub/GenHub.Core/Interfaces/GameClients/IGameClientHashRegistry.cs:
```diff
@@ -0,0 +1,74 @@
+    /// <param name="hash">The SHA-256 hash of the executable.</param>
+    /// <param name="info">The game client information.</param>
+    /// <returns>True if the hash was added, false if it already existed.</returns>
+    bool AddHash(string hash, GameClientInfo info);
```
Maybe this method can be renamed to `TryAddHash`. Then the return type makes a bit more sense (to me). Maybe it's also slightly misleading as there is no `out` param, but I think this might be slightly confusing. `AddedHash`? I'm just thinking out loud, feel free to ignore this entirely.

undead2146 replied:
Done

RMWMStevens commented on GenHub/GenHub.Core/Models/Manifest/ManifestIdGenerator.cs:
```diff
@@ -12,54 +13,103 @@ public static class ManifestIdGenerator
-    /// <returns>A normalized manifest identifier in the form 'schemaVersion.manifestVersion.publisher.content'.</returns>
-    public static string GeneratePublisherContentId(string publisherId, string contentName, int userVersion = 0)
+    /// <param name="suffix">Optional suffix for content type (e.g., '-mod', '-mappack'). Defaults to empty.</param>
+    /// <returns>A normalized manifest identifier in the form 'schemaVersion.manifestVersion.publisher.contentType.content[suffix]'.</returns>
```
Please document exceptions if you throw them intentionally. This way it's clear from the outside that `ArgumentNullException` and `ArgumentException` are to be expected under certain conditions.

undead2146 replied:
Done, added docstring.

RMWMStevens commented on GenHub/GenHub.Core/Models/Manifest/ManifestIdGenerator.cs:
```diff
@@ -12,54 +13,103 @@ public static class ManifestIdGenerator
-    /// <returns>A normalized manifest identifier in the form 'manifestVersion.userVersion.installation.game'.</returns>
-    public static string GenerateGameInstallationId(GameInstallation installation, GameType gameType, int userVersion = 0)
+    /// <param name="userVersion">User-specified version (e.g., "1.08", "1.04", or integer like 0, 1, 2). If null, defaults to 0.</param>
+    /// <param name="suffix">Optional suffix for content type (e.g., '-installation', '-client'). Defaults to '-installation'.</param>
```
You say that it defaults to `-installation`, but I don't see it in the code. Where exactly is the default being set? I see that the suffix is Normalized in a method of the same name, but there an exception is being thrown when the value is null or whitespace. Is this param description wrong, or did you forget to add the default behavior?

undead2146 replied:
I removed the suffix paramater from the manifest id's.

RMWMStevens commented on GenHub/GenHub.Core/Models/Manifest/ManifestIdService.cs:
```diff
@@ -1,98 +1,81 @@
+        catch (Exception ex)
         {
-            return OperationResult<ManifestId>.CreateFailure($"Failed to validate manifest ID: {manifestIdString}");
+            return OperationResult<ManifestId>.CreateFailure($"Invalid manifest ID: {ex.Message}");
```
Maybe it'd be nice to add the provided (invalid) manifest ID in the message as well, and _then_ include the Exception message. Might be easier debugging that way.

RMWMStevens commented on GenHub/GenHub.Tests/GenHub.Tests.Core/Features/GameProfiles/ViewModels/GameSettingsViewModelTests.cs:
```diff
@@ -333,6 +333,32 @@ public async Task OnSelectedGameTypeChanged_Should_LoadSettings_WhenNotInitializ
+        var initTask = _viewModel.InitializeForProfileAsync("test", profile);
+
+        // Complete initialization
+        await initTask;
```
Why not just combine line 353 and 356? Is there a reason to do it in two steps?

undead2146 replied:
Done

RMWMStevens commented on GenHub/GenHub.Tests/GenHub.Tests.Core/Models/Manifest/ManifestIdGeneratorTests.cs:
```diff
@@ -317,14 +317,14 @@ public void GeneratePublisherContentId_WithEmptyNormalizedStrings_StillValid()
 
         // Assert
-        Assert.Equal("1.0.test.unknown", result); // Should handle empty segments gracefully with placeholder
+        Assert.Equal("1.0.test.mod.unknown", result); // Should handle empty segments gracefully with placeholder
```
The comment says `Should handle empty segments gracefully with placeholder`, but the `contentName` is no longer set to a value that will return as empty. Should you add a second test for `contentName = "!!!"` _with_ this comment, or remove it from this current tests as `contentName` will no longer return as empty?

RMWMStevens commented on GenHub/GenHub/Features/GameClients/GameClientDetector.cs:
```diff
@@ -0,0 +1,557 @@
+        IEnumerable<GameInstallation> installations,
+        CancellationToken cancellationToken = default)
+    {
+        var sw = System.Diagnostics.Stopwatch.StartNew();
```
I'm being extremely nitpicky again, I know that, but in the previous file you had a `var stopwatch = Stopwatch.StartNew();` (so with the using, and the name was `stopwatch`). Here you don't have the using, and the name is `sw`. It doesn't matter _at all_ for the end result, I just think it'd be neater to have these things consistent, no matter what it's about.

RMWMStevens commented on GenHub/GenHub/Features/GameClients/GameClientDetector.cs:
```diff
@@ -0,0 +1,557 @@
+            var zeroHourVersion = _hashRegistry.GetVersionFromHash(hash, GameType.ZeroHour);
+            GameType detectedGameType;
+            string detectedVersion;
+            if (generalsVersion != "Unknown")
```
Is the capitalization always `Unknown`, or should you do an ignore-case compare instead?

RMWMStevens commented on GenHub/GenHub/Features/Manifest/ContentManifestBuilder.cs:
```diff
@@ -32,11 +37,11 @@ public class ContentManifestBuilder(
         // Build a temporary GameInstallation to generate the ID
         var tempInstallation = new GameInstallation(string.Empty, installType);
-        tempInstallation.HasZeroHour = gameType == GameType.ZeroHour;
+        tempInstallation.SetPaths(null, gameType == GameType.ZeroHour ? "dummy" : null);
```
What does `dummy` do/mean, and should this be documented?

RMWMStevens commented on GenHub/GenHub/Features/Manifest/ManifestGenerationService.cs:
```diff
@@ -1,321 +1,863 @@
+            var builderLogger = NullLogger<ContentManifestBuilder>.Instance;
+            var builder = new ContentManifestBuilder(builderLogger, _hashProvider, _manifestIdService)
+                .WithBasicInfo(publisherId, gameName, manifestVersion.ToString())
+                .WithContentType(ContentType.GameClient, GameType.Generals); // TODO: Determine game type dynamically
```
Should this `TODO` comment be addressed before merging the PR?

undead2146 replied:
I removed the old overload of `CreateGameClientManifestAsync` (the one without a `GameType` parameter) since the updated method signature includes `GameType` explicitly.

RMWMStevens commented on GenHub/GenHub/Features/Manifest/ManifestGenerationService.cs:
```diff
@@ -1,321 +1,863 @@
-        await JsonSerializer.SerializeAsync(stream, manifest, JsonOptions);
+                // Report progress every 50 files
+                _fileCount++;
+                if (_fileCount - _lastReportedCount >= 50)
```
You could do something similar by writing it this way:
`if (_fileCount % 50 == 0)`
Then you don't have to keep that `_lastReportedCount` variable, unless you want to share it between methods. But I already saw that in the `AddGameFilesToManifest` it gets reset, for example.

RMWMStevens commented on GenHub/GenHub/Features/Workspace/Strategies/HardLinkStrategy.cs:
```diff
@@ -54,8 +54,8 @@ public override long EstimateDiskUsage(WorkspaceConfiguration configuration)
-        }
-        else
+    }
+    else
```
This seems like an unintended change. Should you tab both these lines one more time so it's back in line?

RMWMStevens commented on docs/dev/constants.md:
```diff
@@ -189,20 +192,382 @@ Constants related to manifest ID generation, validation, and file operations.
+
+| Constant | Value | Description |
+|----------|-------|-------------|
+| `ManifestIdGenerationTimeoutMs` | `5000` | Timeout for manifest ID generation operations (milliseconds) |
```
Pro tip: If you want these actual files to look nice as well, instead of only the 'result' of this, consider using the `Markdown All In One` extension for VS Code. In Markdown files, you can press Ctrl+Shift+F and it'll format tables in a 'readable' format as well. I use it for work all the time, and I really like the output. Takes you 0 effort to format your MD files, especially tables.

undead2146 commented on GenHub/GenHub.Core/Models/GameClients/GameClient.cs:
```diff
@@ -23,7 +23,7 @@ public class GameClient
     /// <summary>
     /// Gets or sets the base installation ID for this game client (for test compatibility).
```
This is an outdated comment.

undead2146 commented on GenHub/GenHub.Core/Models/GameInstallations/GameInstallation.cs:
```diff
@@ -69,34 +69,122 @@ public GameInstallation(
-                HasGenerals = true;
-                GeneralsPath = generalsPath;
-                _logger?.LogDebug("Found Generals installation at {GeneralsPath}", GeneralsPath);
+                var generalsExe = Path.Combine(generalsPath, "generals.exe");
```
Correct, done.

undead2146 commented on GenHub/GenHub.Core/Models/Manifest/ManifestIdGenerator.cs:
```diff
@@ -12,54 +13,103 @@ public static class ManifestIdGenerator
-            throw new ArgumentException("Publisher ID cannot be empty", nameof(publisherId));
-        if (string.IsNullOrWhiteSpace(contentName))
-            throw new ArgumentException("Content name cannot be empty", nameof(contentName));
+        if (publisherId == null)
```
Done, reverted this change.

undead2146 commented on GenHub/GenHub.Core/Models/Manifest/ManifestIdGenerator.cs:
```diff
@@ -81,26 +131,32 @@ private static string Normalize(string input)
+            ContentType.MapPack => "mappack",
+            ContentType.LanguagePack => "languagepack",
+            ContentType.ContentBundle => "contentbundle",
+            ContentType.PublisherReferral => "publisherreferral",
```
Done, this makes more sense.

undead2146 commented on GenHub/GenHub.Tests/GenHub.Tests.Core/Features/Manifest/ContentManifestBuilderTests.cs:
```diff
@@ -51,15 +51,19 @@ public ContentManifestBuilderTests()
-        _manifestIdServiceMock.Setup(x => x.GeneratePublisherContentId(It.IsAny<string>(), It.IsAny<string>(), It.IsAny<int>()))
-            .Returns((string p, string c, int v) =>
+        _manifestIdServiceMock.Setup(x => x.GeneratePublisherContentId(It.IsAny<string>(), It.IsAny<GenHub.Core.Models.Enums.ContentType>(), It.IsAny<string>(), It.IsAny<int>()))
+            .Returns((string p, GenHub.Core.Models.Enums.ContentType ct, string c, int v) =>
```
Done

undead2146 commented on GenHub/GenHub/Features/GameClients/GameClientDetector.cs:
```diff
@@ -0,0 +1,557 @@
+        if (!Directory.Exists(path))
+        {
+            sw.Stop();
+            await Task.Delay(1, cancellationToken);
```
Removed the random delay.

undead2146 commented on GenHub/GenHub/Features/Manifest/ManifestGenerationService.cs:
```diff
@@ -1,321 +1,863 @@
+    IFileHashProvider hashProvider,
+    IManifestIdService manifestIdService) : IManifestGenerationService
 {
+    private readonly ILogger<ManifestGenerationService> _logger = logger;
```
Done, removed readonly since the fields aren't mutated
